### ISSUE TITLE
DataStore ObserveQuery's completion handler should only fire twice on first initialization

### ISSUE THREAD

**Issue by ethan021021** [aws-amplify/amplify-swift#2575]
### Describe the bug

Hello,

When using datastore's `observeQuery` on a fresh app install I get the following on app initialization (I'm observing the `Member` query on app init to get the ID of the user):

```
"[debug] snapshot [] isSynced: false"

"[debug] snapshot [] isSynced: true"

"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: true"
```

Is this the correct behavior? I'd expect `observeQuery` to fire the completion handler twice and further on when the `Member` model updates. I'm currently not updating the `Member` model at all here only listening at this point in time.

Now, when I launch the app and start observing  _after_ the `Member` model has been successfully synced to the SQLite DB from the previous app launch I see the following:

```
"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: false"

"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: true"
```

^ This is the exact behavior I require as I'm trying to use the `isSynced` property to determine when to make additional API calls, hide loading indicators, and run analytics _only_ once. Is there a current workaround for this? Is this intended behavior?

Thanks

### Steps To Reproduce

```swift
Steps to reproduce the behavior:
1. Install the app on a fresh device (or delete the app and reinstall it)
2. ObserveQuery on any Model and run the app (incorrect behavior happens here)
3. Relaunch the app after the first app run and notice that the correct behavior is happening
```

### Expected behavior

I'd expect the `ObserveQuery` completion to only fire twice:

1. One fire for isSynced: false (which means syncing is currently in progress)
2. One fire for isSynced: true (when syncing is finished)

### Amplify Framework Version

2.1.0

### Amplify Categories

DataStore

### Dependency manager

Swift PM

### Swift version

5.6

### CLI version

10.4.0

### Xcode version

14.1

### Relevant log output

Incorrect behavior:

```
"[debug] snapshot [] isSynced: false"

"[debug] snapshot [] isSynced: true"

"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: true"
```

Correct behavior:

```
"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: false"

"[debug] snapshot [HAServiceKit.Member(id: "REDACTED", username: Optional("test2"), circle: Optional("purple"), meetingsJoined: Optional(473), deviceToken: Optional("REDACTED"), createdAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-08-24 19:50:47 +0000)), updatedAt: Optional(Amplify.Temporal.DateTime(foundationDate: 2022-11-15 23:10:31 +0000)))] isSynced: true"
```

### Is this a regression?

No

### Regression additional context

I'm not sure if this is a regression or not as I'm unsure if it was fixed in 1.28

### Device

iPhone 14 Pro

### iOS Version

iOS 15.0

**Comment by ethan021021** [comment]
Any progress on this issue? Thanks 😄

**Comment by ethan021021** [comment]
Hey all, is there still a known issue you're looking into? Thank you!

**Comment by atierian** [comment]
Hi @ethan021021, apologies for the long delay in getting back to you on this. 
We're able to reproduce this and are working towards a fix. We'll update you here with anything new to share.

**Comment by atierian** [comment]
@ethan021021 we've identified a fix for this. It's currently in the branch [5d/fix-model-synced](https://github.com/aws-amplify/amplify-swift/tree/5d/fix-model-synced). We're planning on releasing this tomorrow morning and encourage you to try it out by targeting the branch in the meantime. If you experience any issues with it, please let us know asap so we can look to address them.

Thanks again for your patience here!

**Comment by ethan021021** [comment]
Hi @atierian ! I've just tested out the `5d/fix-model-synced` branch and it's working exactly as expected now. I appreciate the push to implement this fix! Thanks a bunch!

**Comment by ethan021021** [comment]
@atierian Sorry I misspoke,

This issue has been fixed if the DataStore DB has data already in it. But the issue persists if you clear the DataStore, and then observeQuery on a Model with an empty/fresh DataStore.

The results when Data is already in the DS:
```
"[debug] isSynced false"
"[debug] isSynced true"
```

The results when there is no records in the DS:
```
"[debug] isSynced false"
"[debug] isSynced true"
"[debug] isSynced true"
```

Would you mind taking another look? Thanks so much

**Comment by atierian** [comment]
That first snapshot where `isSynced` is true should contain all of the records. Is that not the case?
```
"[debug] isSynced false"
"[debug] isSynced true" // <-- this shouldn't be empty anymore.
"[debug] isSynced true"
```

**Comment by ethan021021** [comment]
The first snapshot where `isSynced` is true does contain all the records!

Is printing `isSynced true` twice in this situation expected behavior when DataStore is initializing for the first time? Maybe I'm misunderstand something

Here's the results I'm seeing:

```
"[debug] record count 0 isSynced false"
"[debug] record count 16 isSynced true"
"[debug] record count 16 isSynced true"
```

**Comment by atierian** [comment]
> Is printing isSynced true twice in this situation expected behavior when DataStore is initializing for the first time?

It is not. You should expect new snapshots on mutations, but two snapshots immediately following each other should not be identical. We've addressed this in the same branch, you can resolve to the latest commit and try it out now.

**Comment by ethan021021** [comment]
> > Is printing isSynced true twice in this situation expected behavior when DataStore is initializing for the first time?
> 
> It is not. You should expect new snapshots on mutations, but two snapshots immediately following each other should not be identical. We've addressed this in the same branch, you can resolve to the latest commit and try it out now.

The new commit works wonderfully! Thanks so much. 

I saw some tests were added to check against this specific issue, do the test's cover both cases regarding DS having data already in it/DS being empty?

**Comment by atierian** [comment]
> I saw some tests were added to check against this specific issue, do the test's cover both cases regarding DS having data already in it/DS being empty?

That's correct. There were already test cases covering DataStore having data, this fix also includes additional tests covering the _DataStore is empty_ scenario.

**Comment by ethan021021** [comment]
That's great. Thank you!

**Comment by ethan021021** [comment]
Any idea on if this fix will be merged today or tomorrow? Would love to include it in our next release. Thanks!

**Comment by atierian** [comment]
Thanks for your patience here. We decided to do some more regression testing on this before releasing. At this point, all signs indicate that it will be released at some point tomorrow. I'll keep you updated.

**Comment by atierian** [comment]
Hey @ethan021021. The fix has been released in [v2.11.3](https://github.com/aws-amplify/amplify-swift/releases/tag/2.11.3). 
Please let us know if this release resolves your issue or if you have any problems with it. Thanks!
